plugins/utils/auto-save.nix

```diff
@@ -73,72 +102,57 @@ in
-              return true -- met condition(s), can save
+            if utils.not_in(fn.getbufvar(buf, "&filetype"), {}) then
+              return true
             end
-            return false -- can't save
+            return false
           end
         ''
         ''
           Function that determines whether to save the current buffer or not.
           - return true: if buffer is ok to be saved
           - return false: if it's not ok to be saved
+
+          In this example, the second argument of `utils.not_in(..., {})`
+          determines which filetypes will be ignored by auto-save plugin.
+
+          Buffers that are `nomodifiable` are not saved by default.
         '';
 
-    writeAllBuffers = helpers.defaultNullOpts.mkBool false ''
+    write_all_buffers = helpers.defaultNullOpts.mkBool false ''
       Write all buffers when the current one meets `condition`.
     '';
 
-    debounceDelay = helpers.defaultNullOpts.mkInt 135 ''
+    noautocmd = helpers.defaultNullOpts.mkBool false ''
+      Do not execute autocmds when saving.
+    '';
+
+    lockmarks = helpers.defaultNullOpts.mkBool false ''
+      Lock marks when saving, see `:h lockmarks` for more details.
+    '';
+
+    debounce_delay = helpers.defaultNullOpts.mkUnsignedInt 1000 ''
       Saves the file at most every `debounce_delay` milliseconds.
     '';
 
-    callbacks =
-      mapAttrs (name: desc: helpers.mkNullOrLuaFn "The code of the function that runs ${desc}.")
-        {
-          enabling = "when enabling auto-save";
-          disabling = "when disabling auto-save";
-          beforeAssertingSave = "before checking `condition`";
-          beforeSaving = "before doing the actual save";
-          afterSaving = "after doing the actual save";
-        };
+    debug = helpers.defaultNullOpts.mkBool false ''
+      Log debug messages to `auto-save.log` file in NeoVim cache directory.
+    '';
   };
 
-  config =
-    let
-      options = {
-        enabled = cfg.enableAutoSave;
-        execution_message = with cfg.executionMessage; {
-          inherit message dim;
-          cleaning_interval = cleaningInterval;
-        };
-        trigger_events = cfg.triggerEvents;
-        inherit (cfg) condition;
-        write_all_buffers = cfg.writeAllBuffers;
-        debounce_delay = cfg.debounceDelay;
-        callbacks = with cfg.callbacks; {
-          inherit enabling disabling;
-          before_asserting_save = beforeAssertingSave;
-          before_saving = beforeSaving;
-          after_saving = afterSaving;
-        };
-      } // cfg.extraOptions;
-    in
-    mkIf cfg.enable {
-      extraPlugins = [ cfg.package ];
-
-      extraConfigLua = ''
-        require('auto-save').setup(${helpers.toLuaObject options})
-      '';
-
-      keymaps =
-        with cfg.keymaps;
-        optional (toggle != null) {
-          mode = "n";
-          key = toggle;
-          action = ":ASToggle<CR>";
-          options.silent = cfg.keymaps.silent;
-        };
```

Comment:
This could be kept, using an `extraConfig` attribute:

```nix
extraConfig = cfg: {
  keymaps = mkIf (toggle != null) {
    mode = "n";
    key = cfg.keymaps.toggle;
    action = ":ASToggle<CR>";
    options.silent = cfg.keymaps.silent;
  };
};
```

(Note: prefer `mkIf` over `optional` for config definitions)

Reply:
As for keymaps, there is a [section in the plugin repo](https://github.com/okuuva/auto-save.nvim/?tab=readme-ov-file#-usage) that describes how to add keymap. But it is not default or anything so I decided to remove them.

Reply:
If you don't want to keep the `keymaps` option, you should add a [mkRemovedOptionModule](https://noogle.dev/f/lib/mkRemovedOptionModule) assertion to `imports`, with migration instructions.

Something along these lines:

```nix
(mkRemovedOptionModule (basePluginPath ++ [ "keymaps" ]) ''
  Use the top-level `keymaps` option to create a keymap that runs :ASToggle

  keymaps = [
    { key = "<leader>s"; action = "<cmd>ASToggle<CR>"; }
  ];
'')
```

(note the warning is printed as stderr, so no fancy markdown formatting)

Reply:
On a technical level, though, we _can_ continue to support the option (to avoid breaking people's configs).

`extraOptions` sit alongside the settings options; unlike settings, these don't have to map directly to upstream's options and can be anything we think is useful enough to continue maintaining.

I do agree that this particular keymap is pretty basic though, and we probably wouldn't add it if it were a new plugin. So up to you.

Reply:
> I do agree that this particular keymap is pretty basic though, and we probably wouldn't add it if it were a new plugin.

Yep, I think it should be dropped for consistency with other/new plugins.